backends/ent/store.go

```diff
@@ -0,0 +1,479 @@
+// --------------------------------------------------------------
+// SPDX-FileCopyrightText: Copyright © 2024 The Protobom Authors
+// SPDX-FileType: SOURCE
+// SPDX-License-Identifier: Apache-2.0
+// --------------------------------------------------------------
+package ent
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/protobom/protobom/pkg/sbom"
+	"github.com/protobom/protobom/pkg/storage"
+
+	"github.com/protobom/storage/internal/backends/ent"
+	"github.com/protobom/storage/internal/backends/ent/document"
+	"github.com/protobom/storage/internal/backends/ent/documenttype"
+	"github.com/protobom/storage/internal/backends/ent/externalreference"
+	"github.com/protobom/storage/internal/backends/ent/hashesentry"
+	"github.com/protobom/storage/internal/backends/ent/identifiersentry"
+	"github.com/protobom/storage/internal/backends/ent/metadata"
+	"github.com/protobom/storage/internal/backends/ent/node"
+	"github.com/protobom/storage/internal/backends/ent/nodelist"
+	"github.com/protobom/storage/internal/backends/ent/purpose"
+)
+
+var errInvalidEntOptions = errors.New("invalid ent backend options")
+
+// Store implements the storage.Storer interface.
+func (backend *Backend) Store(doc *sbom.Document, opts *storage.StoreOptions) error {
+	if backend.BackendOptions == nil {
+		backend.BackendOptions = NewBackendOptions()
+	}
+
+	if opts == nil {
+		opts = &storage.StoreOptions{
+			BackendOptions: backend.BackendOptions,
+		}
+	}
+
+	if _, ok := opts.BackendOptions.(*BackendOptions); !ok {
+		return fmt.Errorf("%w", errInvalidEntOptions)
+	}
+
+	backend.init(backend.BackendOptions)
+
+	defer backend.client.Close()
+
+	if err := backend.StoreMetadata(doc.Metadata); err != nil {
+		return fmt.Errorf("%w", err)
```

Comment:
nit: Add some context to the wrapped error here, or if not just `return err`
```suggestion
		return fmt.Errorf("storing metadata: %w", err)
```

Reply:
I would prefer to just `return err` I guess, but the linter settings require that errors are wrapped when returned. The output is really ugly if I add fail context to each statement that gets bubbled up, so I was attempting to only add context to the error for the final result of each `Store*()` operation

Reply:
I was wrong, the linter seems ok with just `return err`. I'll update